File: src/validators/project.ts
import { ValidationResult, ValidationIssue } from './index.js';
import { promises as fs } from 'fs';
import path from 'path';

export async function validateProject(
  projectPath: string,
  platform?: 'ios' | 'android',
  lane?: string
): Promise<ValidationResult> {
  const issues: ValidationIssue[] = [];

  // Check project path exists
  try {
    const stats = await fs.stat(projectPath);
    if (!stats.isDirectory()) {
      issues.push({
        level: 'error',
        code: 'PROJECT_NOT_DIRECTORY',
        message: `Project path is not a directory: ${projectPath}`,
        suggestion: 'Provide a path to your project directory',
      });
      return { valid: false, issues };
    }
  } catch {
    issues.push({
      level: 'error',
      code: 'PROJECT_NOT_FOUND',
      message: `Project path does not exist: ${projectPath}`,
      suggestion: 'Verify the project path is correct',
    });
    return { valid: false, issues };
  }

  // Check platform directories and Fastfiles
  const platforms = platform ? [platform] : ['ios', 'android'];

  for (const plat of platforms) {
    const platformDir = path.join(projectPath, plat);
    const fastfilePath = path.join(platformDir, 'fastlane', 'Fastfile');

    try {
      await fs.access(platformDir);
    } catch {
      if (platform === plat) {
        issues.push({
          level: 'error',
          code: 'PLATFORM_DIR_NOT_FOUND',
          message: `Platform directory not found: ${platformDir}`,
          suggestion: `Create the ${plat} directory in your project`,
        });
      }
      continue;
    }

    try {
      await fs.access(fastfilePath);
    } catch {
      issues.push({
        level: platform === plat ? 'error' : 'warning',
        code: 'FASTFILE_NOT_FOUND',
        message: `Fastfile not found: ${fastfilePath}`,
        suggestion: `Run 'fastlane init' in the ${plat} directory`,
      });
      continue;
    }

    // If a specific lane was requested, check it exists
    if (lane && platform === plat) {
      const lanes = await extractLaneNames(fastfilePath);
      if (!lanes.includes(lane)) {
        issues.push({
          level: 'error',
          code: 'LANE_NOT_FOUND',
          message: `Lane '${lane}' not found in ${plat}/fastlane/Fastfile`,
          suggestion: `Available lanes: ${lanes.join(', ') || 'none found'}`,
        });
      }
    }
  }

  return {
    valid: !issues.some(i => i.level === 'error'),
    issues,
  };
}

async function extractLaneNames(fastfilePath: string): Promise<string[]> {
  try {
    const content = await fs.readFile(fastfilePath, 'utf-8');
    const laneMatches = content.matchAll(/^\s*lane\s+:(\w+)/gm);
    return Array.from(laneMatches, m => m[1]);
  } catch {
    return [];
  }
}

File: src/fastlane_mcp/validators/project.py
"""Project structure validation."""

from pathlib import Path
from fastlane_mcp.validators.types import IssueLevel, ValidationIssue
from fastlane_mcp.utils.paths import find_fastlane_dir


async def validate_project(
    project_path: str,
    platform: str | None,
    lane: str | None
) -> list[ValidationIssue]:
    """Validate project structure and lane existence.

    Args:
        project_path: Path to the project root
        platform: Platform (ios, android, or None for shared)
        lane: Lane name to validate (optional)

    Returns:
        List of validation issues found
    """
    issues: list[ValidationIssue] = []
    project = Path(project_path)

    # Find fastlane directory (checks both platform-specific and root)
    fastlane_dir = find_fastlane_dir(project, platform)

    if fastlane_dir is None:
        # Build helpful error message showing what was checked
        if platform:
            issues.append(ValidationIssue(
                level=IssueLevel.ERROR,
                code="NO_FASTFILE",
                message=(
                    f"Fastfile not found at {project / platform / 'fastlane' / 'Fastfile'} "
                    f"or {project / 'fastlane' / 'Fastfile'}"
                ),
                suggestion="Run 'fastlane init' to create a Fastfile"
            ))
        else:
            issues.append(ValidationIssue(
                level=IssueLevel.ERROR,
                code="NO_FASTFILE",
                message=f"Fastfile not found at {project / 'fastlane' / 'Fastfile'}",
                suggestion="Run 'fastlane init' to create a Fastfile"
            ))
        return issues  # Can't validate lane without Fastfile

    fastfile = fastlane_dir / "Fastfile"

    # Validate lane exists if specified
    if lane:
        content = fastfile.read_text()
        # Simple check for lane definition
        if f"lane :{lane}" not in content and f"private_lane :{lane}" not in content:
            issues.append(ValidationIssue(
                level=IssueLevel.WARNING,
                code="LANE_NOT_FOUND",
                message=f"Lane '{lane}' not found in Fastfile",
                suggestion=f"Available lanes can be listed with 'fastlane lanes'"
            ))

    return issues
